server/grpc_service.go

```diff
@@ -234,66 +234,90 @@ func (s *Server) StoreHeartbeat(ctx context.Context, request *pdpb.StoreHeartbea
 }
 
 // RegionHeartbeat implements gRPC PDServer.
-func (s *Server) RegionHeartbeat(ctx context.Context, request *pdpb.RegionHeartbeatRequest) (*pdpb.RegionHeartbeatResponse, error) {
-	if err := s.validateRequest(request.GetHeader()); err != nil {
-		return nil, errors.Trace(err)
-	}
+func (s *Server) RegionHeartbeat(stream pdpb.PD_RegionHeartbeatServer) error {
+	for {
+		request, err := stream.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
 
-	cluster := s.GetRaftCluster()
-	if cluster == nil {
-		return &pdpb.RegionHeartbeatResponse{Header: s.notBootstrappedHeader()}, nil
-	}
+		// TODO: should we check headers here?
 
-	resp := &pdpb.RegionHeartbeatResponse{}
-	region := newRegionInfo(request.GetRegion(), request.GetLeader())
-	region.DownPeers = request.GetDownPeers()
-	region.PendingPeers = request.GetPendingPeers()
-	if region.GetId() == 0 {
-		pberr := &pdpb.Error{
-			Type:    pdpb.ErrorType_UNKNOWN,
-			Message: fmt.Sprintf("invalid request region, %v", request),
+		resp := &pdpb.RegionHeartbeatResponse{}
```

Comment:
As I said before, please extract a function which returns `(pdpb.RegionHeartbeatResponse, error)`, then send the response.